File: backend/src/agents/baseAgent.ts
/** @format */

import { Configuration, OpenAIApi } from "openai";

interface AgentConfig {
	name: string;
	description: string;
}

export class BaseAgent {
	private openai: OpenAIApi;
	public name: string;
	public description: string;

	constructor(config: AgentConfig) {
		this.name = config.name;
		this.description = config.description;

		// Initialize OpenAI API
		const configuration = new Configuration({
			apiKey: process.env.OPENAI_API_KEY,
		});
		this.openai = new OpenAIApi(configuration);
	}

	/**
	 * Method to handle queries for this agent
	 * @param input - The input string to process
	 * @returns - The output from OpenAI
	 */
	async handleQuery(input: string): Promise<string> {
		try {
			const response = await this.openai.createChatCompletion({
				model: "gpt-4",
				messages: [
					{ role: "system", content: `Agent: ${this.name}. ${this.description}` },
					{ role: "user", content: input },
				],
				max_tokens: 1000,
				temperature: 0.7,
			});

			return response.data.choices[0].message?.content || "No response generated.";
		} catch (error) {
			console.error(`[${this.name} Error]:`, error);
			throw new Error("Failed to process the query.");
		}
	}
}

File: backend/agents/base_agent.py
from backend.interface.agent_interface import AgentInterface
from utils.llm import LLM


class BaseAgent(AgentInterface):
    """
    Base class for all agents. Implements shared functionality.
    """

    def __init__(
        self,
        model_name: str,
        temperature: float,
        dev_msg: str,
        user_msg: str,
        max_tokens: int,
    ):
        """
        Initialize the BaseAgent with an LLM instance.

        Args:
            model_name (str): The name of the LLM model.
            temperature (float): The temperature for LLM responses.
        """
        self.llm = LLM(
            model_name=model_name,
            temperature=temperature,
            dev_msg=dev_msg,
            user_msg=user_msg,
            max_tokens=max_tokens,
        )

    def process(self) -> str:
        """
        Default implementation of the process method.

        Args:
            input_data (str): The input data to process.

        Returns:
            str: The response generated by the LLM.
        """
        raise NotImplementedError("Each agent must implement the `process` method.")